livre/core/pipeline/SimpleExecutor.cpp

```diff
@@ -0,0 +1,160 @@
+
+#include <livre/core/pipeline/Filter.h>
+#include <livre/core/pipeline/Workers.h>
+#include <livre/core/pipeline/Executable.h>
+#include <livre/core/pipeline/FutureMap.h>
+#include <livre/core/pipeline/AsyncData.h>
+#include <livre/core/pipeline/Future.h>
+#include <livre/core/pipeline/Promise.h>
+
+#include <livre/core/data/NodeId.h>
+#include <livre/core/data/LODNode.h>
+#include <livre/core/render/GLContext.h>
+
+#include <lunchbox/mtQueue.h>
+
+#include <boost/thread.hpp>
+
+#include <algorithm>
+#include <iterator>
+
+namespace livre
+{
+
+bool operator<( const Future& future1, const Future& future2 )
+{
+    return future1._impl.get() < future2._impl.get();
+}
+
+struct SimpleExecutor::Impl
+{
+
+    Impl( const size_t threadCount, GLContextPtr glContext )
+        : _workers( threadCount, glContext )
+        , _unlockPromise( DataInfo( "LoopUnlock", getType< bool >( )))
+        , _workThread( boost::thread( boost::bind( &Impl::schedule, this )))
+    {}
+
+    ~Impl()
+    {
+        _mtWorkQueue.clear();
+        _mtWorkQueue.push( 0 );
+        {
+            ScopedLock lock( _promiseReset );
+             _unlockPromise.set( true );
+        }
+        _workThread.join();
+    }
+
+    void schedule()
+    {
+        std::set< Future > inputConditions = { _unlockPromise.getFuture() };
+        Executables executables;
+        while( true )
+        {
+            const Futures inputConds( inputConditions.begin(), inputConditions.end( ));
+            waitForAny( inputConds );
+
+            {
+                ScopedLock lock( _promiseReset );
+                if( _unlockPromise.getFuture( ).isReady( ))
+                {
+                    while( !_mtWorkQueue.empty( ))
+                    {
+                        Executable* exec = _mtWorkQueue.pop();
+                        if( !exec )
+                            return;
+
+                        executables.push_back( exec );
+                    }
+
+                    inputConditions.erase( _unlockPromise.getFuture( ));
+                    _unlockPromise.reset();
+                    inputConditions.insert( _unlockPromise.getFuture( ));
+
+                }
+            }
+
+            Executables::iterator it = executables.begin();
+            while( it != executables.end( ))
+            {
+                Executable* executable = *it;
+                const Futures& preConds = executable->getPreconditions();
+                const FutureMap futureMap( preConds );
+                if( futureMap.isReady( ))
+                {
+                    _workers.schedule( *executable );
+                    it = executables.erase( it );
+                    for( const auto& future: futureMap.getFutures( ))
+                        inputConditions.erase( future );
+                }
+                else
+                {
+                    ++it;
+                    inputConditions.insert( preConds.begin(), preConds.end( ));
```

Comment:
You can do that after line 89, so you don't have to repeat it for all executables every time.

There's also an asymmetry here. You have waitForAll that works on a Future vector, but for isReady you need to create a FutureMap, something that you do from a Future vector. Doesn't it make sense to have a areAllReady function that takes a Futures list?, then you don't need to create the FutureMap.